### PR TITLE
Integration support with Microsoft Teams

### PR DESCRIPTION
**Pull request by Robi9**
https://github.com/rapidpro/rapidpro/issues/1661

### CODE REVIEW

File: handlers/teams/teams.go [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+	"github.com/nyaruka/courier"
+	"github.com/nyaruka/courier/handlers"
+	"github.com/nyaruka/courier/utils"
+	"github.com/nyaruka/gocommon/urns"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	jv                          = JwtTokenValidator{}
```
Comment (norkans7) [inline handlers/teams/teams.go:24]: better be called `jwtTokenValidator` instead of `jv`

File: handlers/teams/teams.go [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+	AllowedSigningAlgorithms    = []string{"RS256", "RS384", "RS512"}
+	ToBotFromChannelTokenIssuer = "https://api.botframework.com"
+	jwks_uri                    = "https://login.botframework.com/v1/.well-known/keys"
```
Comment (norkans7) [inline handlers/teams/teams.go:27]: camel case here `jwtKeysURI`

Comment (norkans7) [reply handlers/teams/teams.go:27]: or `jwksURI` probably

File: handlers/teams/teams.go [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+}
+
+type ConversationAccount struct {
+	ID               string `json:"id"`
+	ConversationType string `json:"conversationType"`
+	TenantID         string `json:"tenantId"`
+	Role             string `json:"role"`
+	Name             string `json:"name"`
+	IsGroup          bool   `json:"isGroup"`
+	AadObjectId      string `json:"aadObjectId"`
```
Comment (norkans7) [inline handlers/teams/teams.go:320]: ID should be in uppercase and all other acronym such as URL too

Comment (norkans7) [reply handlers/teams/teams.go:320]: `ContentURL`, `ServiceURL`, `ChannelID`, ...

File: handlers/teams/teams.go [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+	if payload.Type == "conversationUpdate" {
+		userID := payload.MembersAdded[0].ID
+
+		if userID == "" {
+			return nil, nil
+		}
+
+		act := Activity{}
+
+		act.Text = "Create Conversation"
```
Comment (norkans7) [inline handlers/teams/teams.go:227]: This is going to be saved as a message text?
Is that what we need here?

File: handlers/teams/teams_test.go [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+		mockedCase := tc
+		if mockedCase.Label == "Receive Conversation Update" {
+			mockedCase.Data = strings.Replace(conversationUpdate, "https://smba.trafficmanager.net/br/", newUrl, 1)
+		}
+		casesWithMockedUrls[i] = mockedCase
+	}
+	return casesWithMockedUrls
+}
+
+var defaultSendTestCases = []ChannelSendTestCase{
```
Comment (norkans7) [inline handlers/teams/teams_test.go:260]: Can you add error case test as well?

File: handlers/teams/teams_test.go [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+		Response:          "Handled",
+		Text:              Sp("Hello World"),
+		Attachments:       []string{"https://document-url/foo.pdf"},
+		URN:               Sp("teams:2811:smba.trafficmanager.net/br/"),
+		ExternalID:        Sp("56834"),
+		Date:              Tp(time.Date(2022, 6, 6, 16, 51, 00, 0000000, time.UTC)),
+		Headers:           map[string]string{"Authorization": "Bearer " + access_token},
+		NoQueueErrorCheck: true,
+	},
+	{
```
Comment (norkans7) [inline handlers/teams/teams_test.go:177]: what should the text be in these cases?

**Review by norkans7 — CHANGES_REQUESTED** [review]
You need to update these changes to use the same logger approach as in https://github.com/nyaruka/courier/pull/477/files

**Comment by matmsa27** [comment]
Hi @norkans7 any updates on this review? tks

**Comment by matmsa27** [comment]
Hey, @rowanseymour @norkans7 any news on this PR?

**Pull request closed without merging**